### PR TITLE
Make ObjectStoreScheme public

### PR DESCRIPTION
**Pull request by orf**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/5911.

# Rationale for this change
 
`ObjectStoreScheme::parse` contains non-trivial logic to resolve a given url into an object store scheme. Exposing this is useful, as it allows consumers to set up their own builders as they would like.

# What changes are included in this PR?

This just makes the types public.

# Are there any user-facing changes?

The types become public.

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Thanks @orf  -- this makes sense to me

I think we should wait for @tustvold  to weigh in on any reason it would is important to keep this implementation trait private. 

We also have some duplicated version of this logic in DataFusion so having it in one central place would be quite nice

**Review by alamb — APPROVED** [review]

File: object_store/src/parse.rs [hunk 1]
```diff
@@ -62,7 +62,7 @@ impl ObjectStoreScheme {
     /// Create an [`ObjectStoreScheme`] from the provided [`Url`]
     ///
     /// Returns the [`ObjectStoreScheme`] and the remaining [`Path`]
-    fn parse(url: &Url) -> Result<(Self, Path), Error> {
+    pub fn parse(url: &Url) -> Result<(Self, Path), Error> {
```
Comment (leoyvens) [inline object_store/src/parse.rs:65]: You want to change this to return `super::Error`

File: object_store/src/parse.rs [hunk 2]
```diff
@@ -43,7 +43,7 @@ impl From<Error> for super::Error {
 
 /// Recognises various URL formats, identifying the relevant [`ObjectStore`]
 #[derive(Debug, Eq, PartialEq)]
-enum ObjectStoreScheme {
+pub enum ObjectStoreScheme {
```
Comment (leoyvens) [inline object_store/src/parse.rs:46]: Maybe slap a `#[non_exhaustive]` on this, so that new variants can be added after a potential 1.0 release

**Review by leoyvens — COMMENTED** [review]
I just wished for this, so thanks! I'm not a maintainer but hopefully this is a helpful review.

**Comment by alamb** [comment]
It appears this PR also has a clippy CI failure that we need to fix before merging

**Comment by orf** [comment]
Hey, I will fix these issues up next week. Thanks for the review 😌

**Comment by drmorr0** [comment]
Would it be possible to add `#[derive(Clone)]` on this as well now that it's public?

**Comment by alamb** [comment]
I took the liberty of adding additional docs / examples to this PR and merging up from main

File: object_store/src/parse.rs (hunk 2 above)
Comment (alamb) [reply object_store/src/parse.rs:64]: done in b9ca112cb1f

File: object_store/src/parse.rs (hunk 1 above)
Comment (alamb) [reply object_store/src/parse.rs:103]: fixed in b9ca112cb1f (by making error public not by returning super::Error)

**Comment by alamb** [comment]
I'll plan to merge this in the next few days unless there are any other comments

**Comment by alamb** [comment]
Thanks again @orf  and @leoyvens  and everyone else

**Comment by orf** [comment]
Thank you for merging this, and thank you for fixing it up for me! Sorry I didn’t manage to find the time.

**Pull request merged**